src/plugins/spec/spec.c

```diff
@@ -0,0 +1,842 @@
+/**
+ * @file
+ *
+ * @brief Source for spec plugin
+ *
+ * @copyright BSD License (see doc/COPYING or http://www.libelektra.org)
+ *
+ */
+
+#include "spec.h"
+#include <string.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <fnmatch.h>
+#include <kdbhelper.h>
+#include <kdbease.h>
+#include <kdberrors.h>
+#include <kdbprivate.h>
+
+#include "metafunctions.h"
+
+#define MAX_CHARS_IN_LONG 26
+
+typedef enum{GET, SET}Direction;
+
+typedef enum{ERROR, WARNING, INFO, IGNORE}OnConflict;
+
+typedef enum{ARRAYMEMBER, INVALID, SUBCOUNT, CONFLICT, OUTOFRANGE, MISSING, NAC}Conflict;
+
+typedef struct
+{
+	OnConflict member;
+	OnConflict invalid;
+	OnConflict count;
+	OnConflict conflict;
+	OnConflict range;
+	OnConflict missing;
+}ConflictHandling;
+
+static char *keyNameToMatchingString(const Key *key)
+{
+	uint8_t arrayCount = 0;
+	char *name = strchr(keyName(key), '/');
+	for(char *ptr = name; *ptr != '\0'; ++ptr)
+		if(*ptr == '#')
+			++arrayCount;
+	char *pattern = elektraMalloc(elektraStrLen(name)+arrayCount);
+	char *dst = pattern;
+	for(char *src = (name+1); *src != '\0'; ++src)
+	{
+		if(*src == '_' && *(src-1) == '/' && (*(src+1) == '/' || *(src+1) == '\0'))
+		{
+			*dst++ = '*';
+		}
+		else if(*src == '#' && *(src-1) == '/' && (*(src+1) == '/' || *(src+1) == '\0'))
+		{
+			*dst++ = '#';
+			*dst++ = '*';
+		}
+		else
+		{
+			*dst++ = *src;
+		}
+	}
+	*dst = '\0';
+	return pattern;
+}
+
+static int matchPatternToKey(const char *pattern, const Key *key)
+{
+	return !fnmatch(pattern, (strchr(keyName(key), '/')+1), FNM_NOESCAPE|FNM_PATHNAME);
```

Comment:
Yes exactly, `_` should match one level, and the escaping rules of Elektra should determine what one level is. Escaping rules are explained in src/libs/elektra/keyname.c ("Escaping rules").

Btw. I tried it out: my first, simpler example also is enough to show the problem 
`user/with\/innerslash/abc` does not match `user/_/abc`, but it should (because / is escaped).

Also I am confused why do you use `FNM_NOESCAPE` if you argued, it should not be used? (Because we want default behaviour) Additionally `FNM_NOESCAPE` only refers to the globbing expression, not to escaping path names. Maybe this is the misunderstanding?

Testcases are also useful to check if other fnmatch implementations are conforming, and it would be easier to discuss different cases.